lib/_emerge/depgraph.py

```diff
@@ -5024,10 +5025,40 @@ def _select_files(self, myfiles):
                 #   2) It takes away freedom from the resolver to choose other
                 #      possible expansions when necessary.
                 if "/" in x.split(":")[0]:
+                    atomarg_atom = Atom(x, allow_repo=True)
+
+                    # TODO: Doing the lookup here unfairly penalises the success case
+                    if atomarg_atom.package and not atomarg_atom.cp.startswith("null/"):
+                        found_any_match = any(
```

Comment:
@zmedico Do you have an idea for how to best do this? In particular, I don't want to penalise the success case (so "dev-build/make" having to incur a lookup).

Reply:
Maybe we can do this expansion inside _dep_expand, and possibly only do it after category expansion has failed?

Reply:
OK, pushed a fixup (not squashed it). I tried a few things, including maybe treating this like the old PROVIDE case for virtuals, but this seemed to be the most elegant.

What do you reckon?

Reply:
Seems like _select_package would be better than _iter_match_pkgs_any.

Reply:
Maybe just use _iter_match_pkgs_any to translate the catagory/package-name, and let _select_package handle the rest.

Reply:
Tried to do that now - but worried I've misunderstood what you mean..

Reply:
OH, thank you!

Reply:
OK, in [`2a0f564` (#1234)](https://github.com/gentoo/portage/pull/1234/commits/2a0f564ddbe30ea2a1343589d5fea7ee5617ac9c), it works, but I'm calling `_add_pkg` myself rather than relying on `_select_package` and then the `_add_pkg` later on in the function.

In [`2cbc189` (#1234)](https://github.com/gentoo/portage/pull/1234/commits/2cbc189d871f47ba223c9096df313ab64996d554), I try to use the existing logic more, but that doesn't work.

Reply:
`_select_package` ends up replacing it with an `installed` Package:
```
pkg=<Package ('binary', '/tmp/tmpruf_hub6/', 'dev-build/make-4.4', 'merge', 'None', '225', '0', '1705479700')>
   binary: dev-build/make-4.4::test_repo
installed: dev-build/make-4.4::test_repo
pkg=<Package ('installed', '/tmp/tmpruf_hub6/', 'dev-build/make-4.4', 'nomerge', 'installed')>
```

Reply:
So, _select_package operates on data from _set_args, so you need to call `_set_args` at some point with updated args. This loop isn't a good place to call `_set_args`, since it's supposed to be called before any packages are added to the graph (before _select_package is ever called), and called with all args at once rather than a single arg. Anyway, the tests pass with this diff:
```diff
--- a/lib/_emerge/depgraph.py
+++ b/lib/_emerge/depgraph.py
@@ -5321,8 +5321,12 @@ class depgraph:
 
                         if pkg:
                             atom = Atom(atom.replace(atom.cp, pkg.cp))
+                            arg.atom = dep.atom = atom
+                            arg.pset.replace([atom])
+                            self._set_args([arg])
 
-                            if not self._add_pkg(pkg, dep):
+                            # goto _select_package below
+                            if False and not self._add_pkg(pkg, dep):
                                 writemsg(
                                     f"\n\n!!! Problem resolving dependencies for {arg.arg=}, {arg.atom=}\n",
```

Reply:
For example, you might modify `args` inside `_select_files` in the `for x in myfiles:` loop that creates the `args` list, or just after that loop. Then your transformations will already be present when `args` is passed to `_set_args`.

Reply:
Thanks Zac, I'm going to try get back to this tonight.